src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs

```diff
@@ -45,71 +46,106 @@ public class ExceptionHandlerMiddleware
+                var task = _next(context);
+                if (!task.IsCompletedSuccessfully)
                 {
-                    _logger.ResponseStartedErrorHandler();
-                    throw;
+                    return Awaited(this, context, task);
                 }
 
-                PathString originalPath = context.Request.Path;
-                if (_options.ExceptionHandlingPath.HasValue)
+                return Task.CompletedTask;
+            }
+            catch (Exception exception)
+            {
+                // Get the Exception, but don't continue processing in the catch block as its bad for stack usage.
+                edi = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            return HandleException(context, edi);
+
+            static async Task Awaited(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)
+            {
+                ExceptionDispatchInfo edi = null;
+                try
                 {
-                    context.Request.Path = _options.ExceptionHandlingPath;
+                    await task;
                 }
-                try
+                catch (Exception exception)
                 {
-                    context.Response.Clear();
-                    var exceptionHandlerFeature = new ExceptionHandlerFeature()
-                    {
-                        Error = ex,
-                        Path = originalPath.Value,
-                    };
-                    context.Features.Set<IExceptionHandlerFeature>(exceptionHandlerFeature);
-                    context.Features.Set<IExceptionHandlerPathFeature>(exceptionHandlerFeature);
-                    context.Response.StatusCode = 500;
-                    context.Response.OnStarting(_clearCacheHeadersDelegate, context.Response);
+                    // Get the Exception, but don't continue processing in the catch block as its bad for stack usage.
+                    edi = ExceptionDispatchInfo.Capture(exception);
+                }
 
-                    await _options.ExceptionHandler(context);
+                if (edi != null)
+                {
+                    await middleware.HandleException(context, edi);
+                }
+            }
+        }
 
-                    if (_diagnosticSource.IsEnabled("Microsoft.AspNetCore.Diagnostics.HandledException"))
-                    {
-                        _diagnosticSource.Write("Microsoft.AspNetCore.Diagnostics.HandledException", new { httpContext = context, exception = ex });
-                    }
+        private async Task HandleException(HttpContext context, ExceptionDispatchInfo edi)
+        {
+            _logger.UnhandledException(edi.SourceException);
+            // We can't do anything if the response has already started, just abort.
+            if (context.Response.HasStarted)
+            {
+                _logger.ResponseStartedErrorHandler();
+                edi.Throw();
```

Comment:
Would the stack be any cleaner if we included these first few lines directly in the catch block and re-threw directly?

Reply:
The first catch (sync) *should* be uncommon; as most exceptions *should* be wrapped in a task; so its more about reusing the same code.

For the async am trying to avoid recreating this issue https://github.com/dotnet/roslyn/issues/26567 by leaving the catch prior to doing all the extra work